### PR TITLE
[Feature] MicroDuck skills and hierarchical PPO tutorials

### PR DESCRIPTION
**Pull request by vmoens**
## Nine-skill prior

The low-level recipe trains standing, forward/backward walking, both sidesteps, forward hopping, both turns and hopping in place. It preserves indices 0-5 and appends the three new entries. Each hopping task is sampled three times as often as each other task. `jump_task(speed=0.0)` retains stationary hopping by default and supports forward hopping with velocity tracking.

The head reward uses the calibrated head-frame forward axis and body-relative yaw. The old IMU-to-beak line points about 41 degrees below that axis and encouraged an upward-looking head. Hop launch and rhythm now measure world-vertical velocity, and stationary drift measures world-horizontal velocity; previously, torso tilt could make horizontal travel earn launch reward and make a vertical hop pay drift. Diagnostics report head angles, repeated takeoffs/landings, airtime, height, survival, turning-rate extrema, net drift, maximum excursion and ground speed.

Warm starting expands only the task embedding of the released six-skill actor: turns copy standing and in-place hopping copies the original hopping row. The original rows and actor outputs are preserved exactly before training. The optimizer starts fresh; actor-only exports leave the critic head freshly initialized. Tests cover checkpoint loading and expansion, calibrated head geometry, both turning directions, both hopping modes and tilted-body reward/measurement regressions. The local MicroDuck selection passes 25 tests with one CUDA-only skip; the additional ground-speed regression also passes.

The validated prior and labeled video are published under `priors/nine-skills-20260913/` at immutable Hugging Face revision **`01ebcefca08850231edc0eb428a0151474559a85`**:

- [Checkpoint](https://huggingface.co/torchrl/microduck-skills/blob/01ebcefca08850231edc0eb428a0151474559a85/priors/nine-skills-20260913/walker.ckpt) and [nine-skill video](https://huggingface.co/torchrl/microduck-skills/resolve/01ebcefca08850231edc0eb428a0151474559a85/priors/nine-skills-20260913/videos/skills.mp4).
- [Held-out metrics](https://huggingface.co/torchrl/microduck-skills/blob/01ebcefca08850231edc0eb428a0151474559a85/priors/nine-skills-20260913/skills.json), [acceptance report](https://huggingface.co/torchrl/microduck-skills/blob/01ebcefca08850231edc0eb428a0151474559a85/priors/nine-skills-20260913/acceptance.json), [training configuration and phase history](https://huggingface.co/torchrl/microduck-skills/blob/01ebcefca08850231edc0eb428a0151474559a85/priors/nine-skills-20260913/training.json), and [hash manifest](https://huggingface.co/torchrl/microduck-skills/blob/01ebcefca08850231edc0eb428a0151474559a85/priors/nine-skills-20260913/hashes.json).

Training ran on an Apple-silicon MacBook with 16 native MuJoCo workers. The selected snapshot has 16,629,760 additional transitions from the released six-skill actor. The initial 12M budget was extended to meet the in-place drift target. Metadata records the reward refinements across phases. Held-out seed 20260915, 32 episodes per skill, 500 steps:

| Criterion | Result |
| --- | --- |
| Survival | 288/288 episodes; every skill 100% |
| Standing head pitch/yaw absolute p95 | 1.08 / 0.37 degrees |
| Forward hopping ground speed | 0.285 m/s (target 0.30 +/- 0.10) |
| Hopping airtime: forward / in place | 43.8% / 44.0%; original release 43.3% |
| Repeated hops | Every episode; 55 takeoffs and at least 54 landings on average |
| In-place drift | 0.0436 m/s net; 0.0462 m/s heading-relative ground drift |
| Left / right turning | +1.016 / -0.802 rad/s; correct direction in every episode |

In-place hopping still travels up to 0.445 m in ten seconds; the model card records this and the slower right turn. The labeled video uses an independent seed, fixed camera orientation and side views, and was visually inspected. All ten uploaded files match their local hashes; the downloaded checkpoint loads and the video is accessible. Existing root artifacts and the six-skill navigation / seven-skill football pairings are preserved. The tutorial's nine-skill override trains its own selector and derives its action count from the loaded task library. The integration tutorial and both training-tutorial paths (original six-skill pair and downloaded nine-skill override) pass end to end in docs mode. No extra unit test was added for the action-count assertion; these executions exercise the actual loading, PPO and deployment behavior.

This stacked branch carries #4361 and the already-merged shared-buffer correction #4341 needed by the native parallel training path.

PR stack: `main` ← #4329 (controller deployment) ← #4332 (`PPOTrainer.from_env`) ← **#4330 (these tutorials)**. All branches are in `pytorch/rl`.

Keep MicroDuck's environment guide and add a separate, standalone training tutorial. Readers can explore tasks and rewards without working through PPO, or go directly to training and reuse the published checkpoints. Each page targets a 10–15 minute read and run. The narrative follows the duck learning to move and then choosing skills to reach a goal. Notebook execution and documentation-build notes live in a collapsed section of the examples README.

| Tutorial | What the reader does | Notebook size |
| --- | --- | --- |
| [MicroDuck: tasks, rewards and simulation](https://docs-preview.pytorch.org/pytorch/rl/4330/tutorials/microduck.html) | Build tasks, choose episode tasks, inspect specs/rewards, add a reward, run the supplied gait | 93 executable lines, 10 cells, largest cell 14 lines |
| [MicroDuck: train skills, then compose behaviors](https://docs-preview.pytorch.org/pytorch/rl/4330/tutorials/microduck_skills.html) | Train a walker → load a trained walker → deploy it → train a skill selector → evaluate the published pair | 143 executable lines, 11 cells, largest cell 27 lines |

Preview URLs refresh when the docs job finishes. Both pages link to each other and appear in the documentation index. The original `microduck.py` URL and `microduck_tuto` reference remain the environment guide. The examples README points to both scripts.

**Explore the environment**

The first tutorial restores task presets, reward weights, explicit reset selection, curriculum sampling, fixed tasks across a CPU batch, backend choices, the supplied gait controller and optional recording. Small executable examples show the actual TensorDict interface:

```python
tasks = torch.stack([
    MicroDuckEnv.standing_task(),
    MicroDuckEnv.tracking_task(0.2),
    MicroDuckEnv.sidestep_task(0.15),
    MicroDuckEnv.jump_task(weight=2.0),
])
env = MicroDuckEnv(download=True, backend="mujoco", tasks=tasks)
reset = env.reset(
    TensorDict(task_id=torch.tensor([[2]]), batch_size=[1]),
    set_state=True,
)
# task_id: [1, 1], command: [1, 2], observation: [1, 56]
# Action: 14 joint-target offsets. Task 2 is held for the episode.
```

The custom-reward example owns its registries so rerunning the cell or continuing to checkpoint deployment in the same Python session cannot change the saved task layout:

```python
class HeadingMicroDuck(MicroDuckEnv):
    REWARD_TERMS = MicroDuckEnv.REWARD_TERMS.copy()
    REWARD_PARAMS = MicroDuckEnv.REWARD_PARAMS.copy()

@HeadingMicroDuck.register_reward("heading", weight=0.0, heading_std=0.5)
def heading(features: TensorDictBase, params: TensorDictBase) -> torch.Tensor:
    yaw_rate = features["angular_velocity"][..., 2]
    return torch.exp(-yaw_rate.square() / params["heading_std"].square())

steady = HeadingMicroDuck.tracking_task(
    0.2, reward_weights={"heading": 1.0}, heading_std=0.3,
)
# diagnostics=True exposes "diagnostic_reward_heading" at every step.
```

**Train both levels with PPOTrainer**

The second tutorial keeps the complete training progression together. The shared six-task library covers standing, forward/backward walking, left/right sidestepping, and jumping. The model factory builds a task embedding and GRU shared by a Gaussian actor and value head.

```python
walker, low_critic = make_models(low_env, hidden_size=32, initial_policy_scale=1.0)
low_trainer = PPOTrainer.from_env(
    low_env, actor=walker, critic=low_critic,
    total_frames=1024, frames_per_batch=256, minibatch_size=64,
    sub_traj_len=64,
    gae_kwargs={"group_key": "task_id", "average_gae": True},
    loss_kwargs={"entropy_coeff": 0.01}, num_epochs=4,
)
low_trainer.train()
```

The two recurrent choices stay visible: consecutive training windows and task-wise advantage normalization. Collector, loss, optimizer and sampling setup come from the middle PR.

After this small update, load the published 10M-transition walker. The loader restores its larger 128-unit architecture, frozen weights, original task order and action scale. The task order matters because skill indices select learned embeddings.

```python
checkpoint_path = os.environ.get("MICRODUCK_WALKER_CHECKPOINT") or hf_hub_download(
    repo_id="torchrl/microduck-skills",
    filename="walker.ckpt",
    revision="4191d7d25c4fd58a5c6e6395fcf8217459fdd073",
)
checkpoint = load_checkpoint(checkpoint_path, weights_only=True)
walker, skill_tasks = load_walker(checkpoint)
action_scale = checkpoint["config"]["env"]["action_scale"]
```

`WaypointMicroDuck` adds goal displacement/orientation and rewards progress toward `(0.5, 0.3)` metres. The task defines physics observations, reward and termination; controller execution belongs to the library.

```python
controller = MicroDuckController(
    walker, skill_tasks, group_key=None, reset_key=None, control_period_s=0.02,
)
training_env = ClosedLoopMultiAction.from_env(
    task_env, controller, steps=5, reward_aggregation="sum",
)
# Action becomes "skill": an integer in [0, len(skill_tasks)).
# Each decision executes up to five actions using fresh observations.
# GRU state and gait clock are carried under "_controller".

high_actor, high_critic = make_navigation_models(training_env)
high_trainer = PPOTrainer.from_env(
    training_env, actor=high_actor, critic=high_critic,
    total_frames=512, frames_per_batch=128, minibatch_size=64,
    loss_kwargs={"entropy_coeff": 0.01}, gamma=0.99, num_epochs=4,
)
high_trainer.train()
```

The high-level networks are ordinary 64–64 MLPs. The walker belongs to the environment and is absent from the optimizer. `gamma` discounts one high-level decision, with summed physical-step rewards. Playback then loads the paired navigation checkpoint and plots distance to the goal. A local walker override retains the locally trained selector.

The final section shows the same controller deployment for 5-vs-5 football. Football and the other games belong to the optional [rl-zoo](https://github.com/vmoens/rl-zoo); [#4375](https://github.com/pytorch/rl/pull/4375) updates the tutorial imports and provides the installed-zoo walkthrough. Skills, prior training and this navigation tutorial remain in TorchRL. Later prior PPOTrainer/Hydra migration and sensor recipes stack in [#4378](https://github.com/pytorch/rl/pull/4378) and [#4379](https://github.com/pytorch/rl/pull/4379).

```python
football_training_env = microduck_skill_env(
    football_env, walker, skill_tasks,
    group_key="agents", steps=5, control_period_s=0.02,
)
# 64 matches × 10 players: skill [64, 10], independent state per player.
```

**Watch results without running training or rendering**

The integration page opens with the new nine-skill grid; the training page opens with navigation. Each generated notebook's first cell needs only IPython. HTML pages capture the same browser player, streaming pinned Hugging Face videos; no video binaries are committed or downloaded by the doc build.

```python
from IPython.display import Video

Video(
    url="https://huggingface.co/torchrl/microduck-skills/resolve/"
    "5824a856131009e1ecbc95713677a8363e5f169b/videos/navigation.mp4",
    width=640,
    html_attributes='controls muted loop playsinline preload="metadata" '
    'style="max-width: 100%"',
)
```

**Short docs, full local training**

| Mode | Low-level transitions | High-level decisions | Workers |
| --- | ---: | ---: | ---: |
| Training tutorial / notebook | 1,024 | 512 | 1 |
| Documentation / tutorial CI | 64 | 64 | 1 |
| Full launcher defaults | 10,000,000 | 1,000,000 | 16 |

```bash
python tutorials/sphinx-tutorials/microduck.py
python tutorials/sphinx-tutorials/microduck_skills.py

python -m examples.microduck.train_skills --num-envs 16 \
    --low-level-frames 10000000 --high-level-frames 1000000 \
    --output-dir ~/microduck-training

# Continue with the original walker and resume its high-level stage:
python -m examples.microduck.train_skills --num-envs 16 \
    --walker-checkpoint ~/microduck-training/walker.ckpt \
    --output-dir ~/microduck-training --resume

python -m examples.microduck.train_skills --smoke --output-dir /tmp/microduck-smoke
```

Both tutorials execute top to bottom with no main guards. The full launcher shares task/model definitions, uses a guarded entry point for spawned workers, logs CSV metrics, saves resumable trainer state and evaluates both stages. Tutorial CI installs the needed dependencies, discovers both pages, and smoke-tests the launcher.

**Released six-skill baseline**

- 10,010,624 low-level physical transitions and 1,001,472 high-level decisions with 16 native MuJoCo CPU workers.
- Walker: 48/48 ten-second skill episodes survived. Speed variation, lateral drift and jumping drift remain; the model card records them.
- Navigation: 26/32 arrivals (81.25%), 9.3 cm mean final distance, 42.6 mean decisions. Evaluation uses the fixed `(0.5, 0.3)` m waypoint, a 5 cm arrival radius and at most 100 decisions. This is not an arbitrary-goal or football evaluation.
- [Personal W&B run: curves and videos](https://wandb.ai/vmoens/torchrl-microduck-skills/runs/microduck-macbook-20260911).
- [Published checkpoints, videos, metrics and training metadata](https://huggingface.co/torchrl/microduck-skills).

**Validation**

- Both generated notebooks execute in fresh kernels with full interactive defaults; the saved navigation pair reaches the goal in the tutorial rollout.
- Both tutorials execute in the same process. Rerunning the custom-reward example leaves MicroDuck's original registries unchanged and checkpoint deployment still works.
- Sphinx-Gallery executes both docs-mode examples in about 3 seconds total on the Mac with cached assets and builds HTML with warnings treated as errors.
- Both first video cells execute before TorchRL, MuJoCo or Hugging Face imports. Both rendered players were played and inspected in the browser; cross-links connect the pages.
- The optional recording snippet produces 50 frames and playable notebook HTML.
- Formatting, focused lint and whitespace checks pass.

Earlier implementation validation remains applicable: 19 MicroDuck cases pass (1 CUDA-only case skips locally); 33 middle-PR trainer regressions pass; the launcher was exercised with two native MuJoCo workers and both checkpoint-resume paths. Export/reload reproduces deterministic walker actions and recurrent outputs with a non-default architecture. The generic controller parent passes all 49 MultiAction/closed-loop cases, including partial termination and a PPO update.

### CODE REVIEW

**Comment by vmoens** [comment]
Temporarily closing while the implementation is split into focused prerequisite PRs. This PR will be rebuilt as the tutorial-only tail after those branches are published.

**Comment by vmoens** [comment]
This broad stack has been split into focused, independently reviewable changes:

- #4329: generic low-level controller and closed-loop multi-action transform
- #4392: MicroDuck controller integration, including reset_key forwarding and the real fallen observation
- #4393: evaluate MicroDuck head pose from the supplied state
- #4394: pretrained MicroDuck policy loading and reference documentation
- #4395: forward and in-place hopping tasks
- #4396: trajectory evaluation metrics
- #4388: replay-sampler checkpoint restoration
- #4389: worker and trajectory batch dimensions
- #4390: categorical logging statistics
- #4391: PPOTrainer.from_env
- #4397: focused training launcher
- #4398: environment and reward-design tutorial
- #4399: skills and hierarchical training tutorial

The ParallelEnv shared-buffer portion was already merged independently in #4341. The earlier PPO PR #4332 is superseded by #4391.

The existing #4329 review was applied before the split: the controller state now uses a collector-safe public namespace, its collector test checks current and next state, and the MicroDuck-specific review items are covered in #4392 with native-backend tests. The outdated review threads on #4329 have been resolved.
